### ISSUE TITLE
Outline a structure for Variable Components

### ISSUE THREAD

**Issue by justvanrossum** [BlackFoundryCom/rcjk-tools#1]
## Static font using Static Components

`glyf` table:

  - Glyph:
    - Outline
      --or--
    - list of Components
  - Component:
    - BaseGlyph
    - Transformation

## Variable font using Static Components

`glyf` table:

  - Glyph:
    - Outline
      --or--
    - list of Components
  - Component:
    - BaseGlyph
    - Transformation

`gvar` table:

  - Deltas for outlines
  - Deltas for the *offsets* of components

`fvar` table:

  - no hidden axes

## Main idea of the proposal

- Variable Component sources are regular glyphs, but are unencoded
- They use hidden axes that are not exposed to the user
- Components determine the designspace location for their sources

## Static font using (non-nested) Variable Components

`glyf` table:

  - Glyph:
    - Outline
      --or--
    - list of Components
  - Component:
    - BaseGlyph
    - Transformation
    - ~~Designspace Location~~ --> `VaCo` table

`VaCo` table:

  - Glyph:
    - list of Designspace Location (one for each Component)

`gvar` table:

  - Deltas for outlines for the *sources* for variable components

`fvar` table:

  - *all* axes are *hidden*

## Variable font using Variable Components

`glyf` table:

  - Glyph:
    - Outline
    --or--
    - list of Components
  - Component:
    - BaseGlyph
    - Transformation
    - ~~Designspace Location~~ --> `VaCo` table

`VaCo` table:

  - Glyph:
    - list of Designspace Location (one for each Component)
  - GlyphVariations:
    - deltas for components transformation and Designspace Location (anything not covered by gvar)

`gvar` table:

  - Deltas for outlines
  --or--
  - Deltas for the *offsets* of components

`fvar` table:

  - all axes used for Variable Components are *hidden*

## To be determined: exactly which transform parameters are needed

RoboCJK uses:
  - `x`
  - `y`
  - `rotation`
  - `scalex`
  - `scaley`
  - `rcenterx` (in _component source_ coordinates)
  - `rcentery` (in _component source_ coordinates)

There are two things to consider:
  1. which transform parameters are convenient for editing
  2. which transform parameters are needed for a font renderer to interpolate the good path between two masters

I am somewhat uncomfortable with the current RoboCJK parameters, as they seem a little arbitrary.

My proposal would be this:

- use the existing 2x2 matrix + offset in the `glyf` table
- specify interpolation of the 2x2 through matrix decomposition (interpolate decomposed scale, rotation and skew parmameters)
- specify an additional "center of transformation" (in component source coordinates) in the `VaCo` table

With this, it is possible to exactly replicate GS-CJK at the masters, but the behavior in-between may sometimes be subtly different. I want to do experiments to figure out how much any such differences are, and whether they are show-stoppers or acceptable.

**Comment by davelab6** [comment]
Since this isn't a table from MS (all caps) or Apple (all lowercase) I propose `Cvar` as a mixed case "Component Variation" table. (Obviously this is potentially confused with the existing `cvar` table ;)

**Comment by davelab6** [comment]
> all axes used for Variable Components are hidden

This suggests to me that we will need a 2nd flag (3rd state) for axes that distinguish "advanced user" axes (hidden by default, but useful for some sophisticated/educated users to interact with directly) from "non-user axes" (which require a higher level UI to interact with, or should never be interacted with.)

**Comment by davelab6** [comment]
> I am somewhat uncomfortable with the current RoboCJK parameters

I hope @raphlinus can review these soon too :)

**Comment by davelab6** [comment]
There should be an interface between variable components and OpenType features, such as `smcp` or `onum` or `sups` (as David Berlow proposed in http://variationsguide.typenetwork.com/#votf)

(It seems David's proposal for "Variable Unicode IDs" (http://variationsguide.typenetwork.com/#vuid) is provisioned for already, by simply having glyphs with the relevant variable component that are correctly encoded.)

**Comment by justvanrossum** [comment]
> (It seems David's proposal for "Variable Unicode IDs" (http://variationsguide.typenetwork.com/#vuid) is provisioned for already, by simply having glyphs with the relevant variable component that are correctly encoded.)

Yes, I think that's indeed a logical (and pleasant) consequence of this proposal.
